dingo/core/nn/nsf.py

```diff
@@ -218,21 +218,27 @@ def sample(self, *x, num_samples=1):
         if self.embedding_net is not None:
             x = torchutils.forward_pass_with_unpacked_tuple(self.embedding_net, x)
         if len(x) > 0:
-            return torch.squeeze(self.flow.sample(num_samples, x))
+            sample = self.flow.sample(num_samples)
+            sl = list(range(sample.ndim - 1))
+            return torch.squeeze(sample, dim=sl)
```

Comment:
I'm trying to understand the issue here. Is the original reason for the squeeze to drop the leading dimension of size 1 when `num_samples=1`?  And the problem occurs for the single-detector time network because the output is a single time, leads to dropping the last dimension? If so, a simpler fix could be
```suggestion
            sample = self.flow.sample(num_samples)
            return torch.squeeze(sample, dim=0)
```

Line 221 also looks like an error, no? Don't we need to pass `x` to `flow.sample`?

Reply:
I think you mostly described the issue. The only additional thing is that by default, sample has an extra dimension. So e.g. if you have `num_samples = 50` and a output dimension of 2, the size of `sample` is actually `(50, 1,  2)`. I think this is important during training where the middle dimension may be something other than 1 to account for the batch. So then in general it is better to squeeze all dims except the last one. 

Yes line 221 should also be passing `x` to `flow.sample`. This has been updated in commit `45a2b17`

Reply:
During training, don't we ever only evaluate the `log_prob`?

Reply:
Okay I stepped through the code, and indeed I see the shape you describe. I would have thought the first dimension corresponded to the batch though, since `x` is of shape `(50, 128)`, with `50` the (repeated) batch dimension. So I don't see why `nflows` is inserting the singleton dimension. Anyway, `torch.squeeze` eliminated it.

With a 1D parameter space (e.g., the initialization network for 1 detector) the squeezing would also drop the last dimension, which would be problematic. So your fix resolves this.

Turning to the first dimension (corresponding to the number of samples), the present version (and previous) gives an error when requesting just one sample. This error is of the same nature as above, since the leading dimension is eliminated by the squeeze, which the sampler class does not expect. A fix for this would be to return `torch.squeeze(sample, dim=1)`, if we are convinced that works in all cases.

Thoughts @max-dax @nihargupte-ph ?

Reply:
Looking more closely, the singleton dimension refers to the number of samples requested from `Flow.sample()`. There are two ways in which this is called:
1. For conditional NPE flows and GNPE models, we call this with `num_samples=1` (setting the requested number of samples instead via the leading dimension of the context). In this case, we want to squeeze `dim=1`.
2. For an unconditional flow, we have no context, so we set the number of samples via  `num_samples>1`. In this case, I believe `x` will be of shape `(num_samples, param_dim)`, and we don't want to squeeze anything. In fact, if we did my suggestion of squeezing `dim=1`, we would run into a problem with the unconditional 1D flow for single-detector networks.

I think what we should do then is
```
if sample.ndim == 3:
    sample = torch.squeeze(sample, dim=1)
return sample
```
What do people think?

Reply:
Perhaps we could alternatively make use of the existing `if/else` statement that refers to the presence of the context.

Reply:
Okay, how about this @max-dax @nihargupte-ph 

```
    def sample(self, *x, num_samples=1):
        if self.embedding_net is not None:
            x = torchutils.forward_pass_with_unpacked_tuple(self.embedding_net, x)
        if len(x) > 0:
            sample = self.flow.sample(num_samples, x)
            assert sample.ndim == 3
            if num_samples == 1:
                return torch.squeeze(sample, dim=1)
            else:
                return torch.squeeze(sample, dim=0)
        else:
            # if there is no context, omit the context argument
            return self.flow.sample(num_samples)
```
and similarly for `sample_and_log_prob()`.

Reply:
I haven't tested it so maybe it works. But if you set `dim=(1 if num_samples==1 else 0)` then wouldn't this not drop the middle dimension?

But in any case, if we know the meaning of the last dimension, and we know it should never get squeezed, isn't the most minimal change to just not squeeze it? It keeps the rest of the logic the same.

Reply:
I've taken a new approach to this, completely removing the squeezing from the `FlowWrapper`, since it is hard to ascertain the various use cases from within that class. This makes the `FlowWrapper` methods much more straightforward.

Instead, the squeezing is done in the `_run_sampler()` methods of the `Sampler` and `GNPESampler` classes, after calling `FlowWrapper.sample_and_log_prob()`. These classes know what they are trying to do, so they know which dimension to squeeze.

One additional optimization I made is to pass the data through the embedding network only once per batch for NPE networks, rather than replicating it `num_samples` times. We instead use the `num_samples` argument of `FlowWrapper.sample_and_log_prob()` to set the number of samples. (This can't be done for GNPE networks because each data sample is transformed differently according to the parameters.)

I had to update one unit test so that it continues to pass. I also tested sampling form NPE and GNPE networks, and it works. It would be good to test single detector networks, as well as unconditional flows (e.g., by running dingo_pipe on an event).

Reply:
I checked sampling with a 1D network using dingo_pipe. I am able to sample and train the unconditional flow.